lib/events.js

```diff
@@ -1105,33 +1030,58 @@ function on(emitter, event, options) {
       { once: true });
   }
 
-  return iterator;
+  function errorHandler (err) {
+    error = err;
+    errored = true;
+
+    if (resume) {
+      resume(Promise.reject(err));
+      resume = null;
+    }
+  }
 
   function abortListener() {
     errorHandler(new AbortError(undefined, { cause: signal?.reason }));
   }
 
   function eventHandler(...args) {
-    const promise = ArrayPrototypeShift(unconsumedPromises);
-    if (promise) {
-      promise.resolve(createIterResult(args, false));
+    if (resume) {
+      resume(args);
+      resume = null;
     } else {
-      unconsumedEvents.push(args);
+      queue.push(args);
     }
   }
 
-  function errorHandler(err) {
-    finished = true;
+  return async function * () {
+    try {
+      while (true) {
+        while (!queue.isEmpty()) {
+          if (errored) {
+            throw error;
+          }
+          yield queue.shift();
+        }
 
-    const toError = ArrayPrototypeShift(unconsumedPromises);
+        if (errored) {
+          throw error;
+        }
 
-    if (toError) {
-      toError.reject(err);
-    } else {
-      // The next time we call next()
-      error = err;
-    }
+        yield await new Promise(resolve => {
+          resume = resolve;
+        });
+      }
+    } finally {
+      eventTargetAgnosticRemoveListener(emitter, event, eventHandler);
```

Comment:
So the behavior here is to drop all events when the iterator is returned right?

So if I `.next()` 5 times and then `.return` I would get `{ done: true, value: undefined }` on all 5?

Reply:
yes, I believe so